### DISCUSSION TITLE
Any way to circumvent the problem of unsupported filesystem watchers?

### DISCUSSION THREAD

**Discussion by tiagorvmartins** [gramineproject/gramine#1553] [Q&A] [answered]
I have an application that I am trying to run inside gramine sgx and apparently it relies on filesystem watchers, which I believe it's not supported according to the docs:

```
Monitoring filesystem events (inotify, fanotify)

Gramine does not currently implement inotify and fanotify APIs. 
Gramine could implement them in the future, if need arises.
```

I can see the log debug line:
`(libos_parser.c:1641:warn_unsupported_syscall) [P1:T7:geth] warning: Unsupported system call inotify_init1`

And then further on:
`ERROR[09-13|15:45:13.356] Failed to start filesystem watcher       err="function not implemented"`

Maybe there is another solution to my problem, 
I have been thinking about this solution:

I could use  **_Temporary in-memory-only files_** using the type tmpfs filesystem mount.
However I need to be able to stop the process and restart it from the place where it stopped.
For this, I would need to be able to mount a few files to the container such that it continues processing from the same point.

And if I am correct, using tmpfs I wouldn't be able to extract this files and saved them outside of the container to use them again -
_"tmpfs do not support mmap and each process has its own, non-shared tmpfs (i.e., processes don’t see each other’s files)"_

**Context:**
- its a binary (geth) that when ran it creates a geth.ipc domain socket "file" and I believe this is the one responsible for triggering the filesystem watcher errors, but not entirely sure.

Thank you, let me know if you need any more information.

**Comment by dimakuv** [comment]
> I could use Temporary in-memory-only files using the type tmpfs filesystem mount.

Sounds like a wrong solution for your use case. Temporary files (`tmpfs`) are **ephemeral** -- they only live as long as the SGX enclave lives. As soon as the SGX enclave is terminated (i.e. the Gramine process is terminated), these temporary files are lost forever.

> Maybe there is another solution to my problem

Typically applications have some configuration knobs to disable fancy Linux features. I would assume that Geth has some knob/config to **not use** filesystem watchers?

Otherwise, if Geth insists on using filesystem watchers, you have bad luck, and you'll have to wait until Gramine implements this feature...

**Reply by tiagorvmartins** [reply]
Thank you for the prompt answer @dimakuv 
Let's assume I could maintain the state of my application with normal fs mounts default chroot, and isolate the only file that depends on filesystem watchers to use tmpfs - I believe that could work?

More on this question, lets say the file structure is this:

- /app/
- /app/state-data/ - mounted using fs.mount type chroot
- /app/watcher/ - mounted using tmpfs

One last question: 
Assuming that the above would be feasible, can I feed into gramine-sgx the watcher file from host and mark it inside gramine as tmpfs - (I believe not, but just wanted to confirm). 

I guess, this file would need to be part of the image itself then and marked as tmpfs, correct?
Thank you

**Reply by dimakuv** [reply, marked as answer]
> Assuming that the above would be feasible, can I feed into gramine-sgx the watcher file from host and mark it inside gramine as tmpfs - (I believe not, but just wanted to confirm).

That's not possible. Files marked in Gramine as `tmpfs` are **completely unrelated to the host files**. They are purely in-Gramine-enclave representations, not backed by real files on the host.
